toolchain/check/testdata/facet/validate_rewrite_constraints.carbon

```diff
@@ -0,0 +1,1353 @@
+  // CHECK:STDERR:
+  F(T);
+}
+
+// --- recursive_facet.carbon
+library "[[@TEST_NAME]]";
+
+interface I { let X:! type; }
+interface K(Y:! type) {  }
+
+fn F(T:! I where .X = {.k: K(I where .X = ())}) {}
+
+fn G(T:! I where .X = {.k: K(I where .X = ())}) {
+  F(T);
+}
+
+// --- fail_facet_type_concrete_types_match_blanket_impl_concrete_types.carbon
+library "[[@TEST_NAME]]";
+
+interface A { let X:! type; }
+interface B { let Y:! type; }
+
+interface C(BB:! B) { let AX:! type; let BY:! type; }
+
+impl forall [AA:! A, BB:! B] AA as C(BB) where .AX = () and .BY = {} {}
+
+fn F(AA:! A where .X = (), BB:! B where .Y = {}) {
+  // The types match but there may be a specialization that specifies different
+  // types and would be prefered for a specific `AA` or `BB`.
+  //
+  // CHECK:STDERR: fail_facet_type_concrete_types_match_blanket_impl_concrete_types.carbon:[[@LINE+4]]:3: error: cannot convert type `AA` that implements `A where .(A.X) = ()` into type implementing `C(BB as B) where .(C(BB as B).AX) = () and .(C(BB as B).BY) = {}` [ConversionFailureFacetToFacet]
+  // CHECK:STDERR:   AA as (C(BB) where .AX = () and .BY = {});
+  // CHECK:STDERR:   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  AA as (C(BB) where .AX = () and .BY = {});
+}
+
+// --- fail_facet_type_concrete_types_become_blank_impl_types.carbon
+library "[[@TEST_NAME]]";
+
+interface A { let X:! type; }
+interface B { let Y:! type; }
+
+interface C(BB:! B) { let AX:! type; let BY:! type; }
+
+impl forall [AA:! A, BB:! B] AA as C(BB) where .AX = AA.X and .BY = BB.Y {}
+
+fn F(AA:! A where .X = (), BB:! B where .Y = {}) {
+  // The types match but there may be a specialization that specifies different
+  // types and would be prefered for a specific `AA` or `BB`.
+  //
+  // CHECK:STDERR: fail_facet_type_concrete_types_become_blank_impl_types.carbon:[[@LINE+4]]:3: error: cannot convert type `AA` that implements `A where .(A.X) = ()` into type implementing `C(BB as B) where .(C(BB as B).AX) = () and .(C(BB as B).BY) = {}` [ConversionFailureFacetToFacet]
+  // CHECK:STDERR:   AA as (C(BB) where .AX = () and .BY = {});
+  // CHECK:STDERR:   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  AA as (C(BB) where .AX = () and .BY = {});
+}
+
+// --- facet_type_concrete_types_match_final_blanket_impl_concrete_types.carbon
+library "[[@TEST_NAME]]";
+
+interface A { let X:! type; }
+interface B { let Y:! type; }
+
+interface C(BB:! B) { let AX:! type; let BY:! type; }
+
+final impl forall [AA:! A, BB:! B] AA as C(BB) where .AX = () and .BY = {} {}
+
+fn F(AA:! A where .X = (), BB:! B where .Y = {}) {
+  AA as (C(BB) where .AX = () and .BY = {});
+}
+
+// --- facet_type_concrete_types_become_final_blank_impl_types.carbon
+library "[[@TEST_NAME]]";
+
+interface A { let X:! type; }
+interface B { let Y:! type; }
+
+interface C(BB:! B) { let AX:! type; let BY:! type; }
+
+final impl forall [AA:! A, BB:! B] AA as C(BB) where .AX = AA.X and .BY = BB.Y {}
```

Comment:
Can we get one more variant of this test where the `as` conversion fails because the `.AX` / `.BY` values don't match? I think this set of tests would all still pass if they were ignored entirely by the explicit `as` conversion.

Reply:
Done